esphome/components/sen5x/sen5x.cpp

```diff
@@ -297,38 +325,104 @@ void SEN5XComponent::update() {
-          uint32_t state1 = states[2] << 16 | states[3];
-          if ((uint32_t) std::abs(static_cast<int32_t>(this->voc_baselines_storage_.state0 - state0)) >
-                  MAXIMUM_STORAGE_DIFF ||
-              (uint32_t) std::abs(static_cast<int32_t>(this->voc_baselines_storage_.state1 - state1)) >
-                  MAXIMUM_STORAGE_DIFF) {
-            this->seconds_since_last_store_ = 0;
-            this->voc_baselines_storage_.state0 = state0;
-            this->voc_baselines_storage_.state1 = state1;
-
-            if (this->pref_.save(&this->voc_baselines_storage_)) {
-              ESP_LOGI(TAG, "Stored VOC baseline state0: 0x%04" PRIX32 " ,state1: 0x%04" PRIX32,
-                       this->voc_baselines_storage_.state0, voc_baselines_storage_.state1);
-            } else {
-              ESP_LOGW(TAG, "Could not store VOC baselines");
-            }
+    this->write_voc_baseline_();
+  }
+
+  this->update_measured_values_();
+
+  if (this->get_pm_number_concentration_and_tps_) {
+    // delay the extra reading to allow update_measured_values to complete.
+    this->set_timeout(30, [this]() { this->update_measured_pm_(); });
+  }
+}
+
+void SEN5XComponent::write_voc_baseline_() {
+  if (this->write_command(SEN5X_CMD_VOC_ALGORITHM_STATE)) {
+    // run it a bit later to avoid adding a delay here
+    this->set_timeout(550, [this]() {
+      uint16_t states[4];
+      if (this->read_data(states, 4)) {
+        uint32_t state0 = states[0] << 16 | states[1];
+        uint32_t state1 = states[2] << 16 | states[3];
+        if ((uint32_t) std::abs(static_cast<int32_t>(this->voc_baselines_storage_.state0 - state0)) >
+                MAXIMUM_STORAGE_DIFF ||
+            (uint32_t) std::abs(static_cast<int32_t>(this->voc_baselines_storage_.state1 - state1)) >
+                MAXIMUM_STORAGE_DIFF) {
+          this->seconds_since_last_store_ = 0;
+          this->voc_baselines_storage_.state0 = state0;
+          this->voc_baselines_storage_.state1 = state1;
+
+          if (this->pref_.save(&this->voc_baselines_storage_)) {
+            ESP_LOGI(TAG, "Stored VOC baseline state0: 0x%04" PRIX32 " ,state1: 0x%04" PRIX32,
+                     this->voc_baselines_storage_.state0, voc_baselines_storage_.state1);
+          } else {
+            ESP_LOGW(TAG, "Could not store VOC baselines");
           }
         }
-      });
-    }
+      }
+    });
+  }
+}
+
+void SEN5XComponent::update_measured_pm_() {
+  if (!this->write_command(SEN5X_CMD_READ_PM_MEASUREMENT)) {
+    this->status_set_warning();
+    ESP_LOGD(TAG, "write error read measurement (%d)", this->last_error_);
+    return;
   }
 
+  this->set_timeout(20, [this]() {
+    uint16_t measurements[10];
+
+    if (!this->read_data(measurements, 10)) {
+      this->status_set_warning();
+      ESP_LOGD(TAG, "read data error (%d)", this->last_error_);
+      return;
+    }
+
+    float pm_n_0_5 = measurements[4] / 10.0;
+    if (measurements[4] == 0xFFFF)
+      pm_n_0_5 = NAN;
```

Comment:
I see the old code already does this, but since changes are being made, it would be better to check for invalid value before doing a division =)

Reply:
Also without the final `f`, it will use `double` instead of `float`
```suggestion
    float pm_n_0_5 = measurements[4] / 10.0f;
    if (measurements[4] == 0xFFFF)
      pm_n_0_5 = NAN;
```

Reply:
Good catch! I'll move all of that validation into its own function to reduce all the repetitions